Make Copter submode definitions private to their modes

peterbarker opened the pull request.

rmackay9 commented on ArduCopter/mode_rtl.cpp:
```diff
@@ -33,6 +33,13 @@ void Copter::ModeRTL::restart_without_terrain()
     }
 }
 
+bool Copter::ModeRTL::navigate_home_complete() const
```
can we call this function, "reached_home" to be consistent with "reached_wp_destination" of AC_WPNAV.?

rmackay9 commented on ArduCopter/mode_flip.cpp:
```diff
@@ -33,6 +33,15 @@
 #define FLIP_PITCH_FORWARD  -1      // used to set flip_dir
 
 // FIXME? these should be instance variables?
```
this FIXME comment should probably go below the FlipState

rmackay9 commented on ArduCopter/mode_althold.cpp:
```diff
@@ -21,6 +21,13 @@ bool Copter::ModeAltHold::init(bool ignore_checks)
 // should be called at 100hz or more
 void Copter::ModeAltHold::run()
 {
+    enum AltHoldModeState {
```
can we leave this change out?  I've got a big PR coming that totally changes the althold states

rmackay9 commented on ArduCopter/mode.h:
```diff
@@ -739,6 +754,15 @@ class ModeGuided : public Mode {
+        Guided_PosVel,
+        Guided_Angle,
+    };
+    GuidedMode mode() const { return guided_mode; }
```
I think we should use a more specific name rather than, "mode".  How about get_submode()? or just submode()?

rmackay9 commented:
I've suggested a few changes but after that I think it looks good, txs!

peterbarker commented:
On Fri, 1 Feb 2019, Randy Mackay wrote:

> can we call this function, "reached_home" to be consistent with "reached_wp_destination" of AC_WPNAV.?

Done.

peterbarker commented:
On Fri, 1 Feb 2019, Randy Mackay wrote:

> this FIXME comment should probably go below the FlipState

Should I just fix the FIXME?  I think the mode definitions would move at 
the same time, making the current comment well-placed...

peterbarker commented:
On Fri, 1 Feb 2019, Randy Mackay wrote:

> +    enum AltHoldModeState {
> 
> can we leave this change out? I've got a big PR coming that totally changes the althold states

I could just delay this PR until you get yours in.

Note that several modes look very much like althold - I did some patches 
at one stage which made them inherit from althold, but wasn't sure a 
pure-virtual intermediate class wouldn't be better, so I never PR'd it.

peterbarker commented:
On Fri, 1 Feb 2019, Randy Mackay wrote:

> I think we should use a more specific name rather than, "mode". How about get_submode()? or just submode()?

Definitely!

Consistency amongst all of these would be nice - but I don't think I've 
changed any of the existing names, just changed their scopes.

rmackay9 commented:
fixing the FIXME sounds fine to me..

Ah, I didn't realise mode() was what is was called already.. ok, it can stay for now then.

peterbarker commented:
I've pushed up a fix for the Heli compilation issue.

@bnsgeyer 
The implementation of `check_dynamic_flight` looks rotten to me.  I believe there are several cases where we might expect to be using the full collective range where we are currently not.  One example would be in mode auto in submode rtl.

Perhaps stepping back and creating a method on `Copter::Mode` - perhaps, `landing_expected()` might work.

bnsgeyer commented:
> I've pushed up a fix for the Heli compilation issue.
> 
> @bnsgeyer
> The implementation of `check_dynamic_flight` looks rotten to me. I believe there are several cases where we might expect to be using the full collective range where we are currently not. One example would be in mode auto in submode rtl.
> 
> Perhaps stepping back and creating a method on `Copter::Mode` - perhaps, `landing_expected()` might work.

@peterbarker  I guess I need to understand what your doing here.  This code was written by Rob.  I believe the idea was to keep the non_manual_throttle modes from commanding full down collective while the aircraft is on the ground.  In the past, we have not trusted the land_complete flag and use the dynamic flight flag instead to handle things that involve the aircraft landing or on the ground with rotors turning.
explain to me why you think we might need full collective range for rtl?  Just wondering.   I will have to look at this in more detail to see what's valid.  For now, it looks like your change just restores the code back to how it worked before this PR.  Correct me if I'm wrong.  Thanks,
Bill

peterbarker commented:
On Fri, 1 Feb 2019, Bill Geyer wrote:

> 
> @peterbarker I guess I need to understand what your doing here. This code was written by Rob. I believe the idea was to keep the non_manual_throttle modes
> from commanding full down collective while the aircraft is on the ground. In the past, we have not trusted the land_complete flag and use the dynamic
> flight flag instead to handle things that involve the aircraft landing or on the ground with rotors turning.
> explain to me why you think we might need full collective range for rtl? Just wondering. I will have to look at this in more detail to see what's valid.

Auto incorporates running RTL when in AutoRTL: https://github.com/ardupilot/ardupilot/blob/master/ArduCopter/mode_auto.cpp#L825

RTL in Land_Run mode may put the vehicle on the ground, requiring the 
i-term leaking and collective changes.

Note we already test for being in the Auto_Land submode and consider that 
non-dynamic - just not Auto_RTL and rtl being in the landing phase....

> For now, it looks like your change just restores the code back to how it worked before this PR. Correct me if I'm wrong. Thanks,

Yes, it does indeed retain the current code.  I was narrowing the scope of 
the submodes - considering those to be internal to the mode itself. 
Worked well - apart from this particular piece of code!

I'd really like to eliminate this code poking the submodes directly, or 
we're likely to get the same bug in different ways later.  If we can catch 
the essence of the check in a single callback on the mode that would be 
ideal.  flightmode->in_final_landing_phase() would be the brute-force 
approach.  I would point out that with pilot-land-repositioning you might 
actually be undergoing "dynamic flight" even in this final landing phase!

> Bill

Yours,
-- 
Peter Barker                          |   Programmer,Sysadmin,Geek.
pbarker@barker.dropbear.id.au	      |   You need a bigger hammer.
:: It's a hack! Expect underscores! - Nigel Williams

bnsgeyer commented:
> I'd really like to eliminate this code poking the submodes directly, or we're likely to get the same bug in different ways later. If we can catch the essence of the check in a single callback on the mode that would be ideal. flightmode->in_final_landing_phase() would be the brute-force approach. I would point out that with pilot-land-repositioning you might actually be undergoing "dynamic flight" even in this final landing phase!

@peterbarker What you are proposing sounds like it will work.  I will look at this more this week and see if I can come up with a callback as you suggested.

Regards,
Bill

peterbarker commented:
This is WIP as it apparently will conflict with a body of work yet to be committed.  Also, I'm doing a bit of rework to use scoped enumerations.

peterbarker commented:
Redoing this after the spool changes go in will be faster than fixing the conflicts.  Closing.
